otaclient/app/create_standby/rebuild_mode.py

```diff
@@ -179,89 +117,45 @@ def _process_persistents(self):
                 ):  # NOTE: not equivalent to perinf.path.exists()
                     _copy_tree.copy_with_parents(perinf.path, self.standby_slot_mp)
 
-    def _save_meta(self):
-        """Save metadata to META_FOLDER."""
-        _dst = self.standby_slot_mp / Path(self.META_FOLDER).relative_to("/")
-        _dst.mkdir(parents=True, exist_ok=True)
-
-        logger.info(f"save image meta files to {_dst}")
-        for meta_f in self.metadata.get_img_metafiles():
-            _src = self._recycle_folder / meta_f.file
-            shutil.copy(_src, _dst)
-
     def _process_symlinks(self):
+        self.update_phase_tracker(wrapper.StatusProgressPhase.SYMLINK)
         symlink_inf_fname = self.metadata.symboliclink.file
-        with open(self._recycle_folder / symlink_inf_fname, "r") as f:
+        with open(self._ota_tmp_image_meta_dir / symlink_inf_fname, "r") as f:
             for entry_line in f:
                 SymbolicLinkInf(entry_line).link_at_mount_point(self.standby_slot_mp)
 
     def _process_regulars(self):
         self.update_phase_tracker(wrapper.StatusProgressPhase.REGULAR)
         self._hardlink_register = HardlinkRegister()
 
-        # limitation on on-going tasks
-        _tasks_tracker = SimpleTasksTracker(
-            max_concurrent=self.MAX_CONCURRENT_TASKS,
-            title="process_regulars",
-        )
-
-        # apply delta
         logger.info("start applying delta...")
         with ThreadPoolExecutor(thread_name_prefix="create_standby_slot") as pool:
-            for _hash, _regulars_set in self.delta_bundle.new_delta.items():
-                # interrupt update if _tasks_tracker collects error
-                if e := _tasks_tracker.last_error:
-                    logger.error(f"interrupt update due to {e!r}")
-                    raise e
-
-                fut = pool.submit(
-                    self._apply_reginf_set,
-                    _hash,
-                    _regulars_set,
-                )
-                _tasks_tracker.add_task(fut)
-                fut.add_done_callback(_tasks_tracker.done_callback)
-
-            logger.info(
-                "all process_regulars tasks are dispatched, wait for finishing..."
+            _mapper = RetryTaskMap(
+                self._process_regular,
+                self.delta_bundle.new_delta.items(),
+                title="process_regulars",
+                max_concurrent=cfg.MAX_CONCURRENT_PROCESS_FILE_TASKS,
+                backoff_max=cfg.CREATE_STANDBY_BACKOFF_MAX,
+                backoff_factor=cfg.CREATE_STANDBY_BACKOFF_FACTOR,
+                max_retry=cfg.CREATE_STANDBY_RETRY_MAX,
+                executor=pool,
```

Comment:
It seems that `_process_regular` doesn't download anything, is this understanding correct?
If so, why `_process_regular` is wrapped with RetryTaskMap?

Reply:
`RetryTaskMap` can manage a group of tasks, not only manage downloading. It is a generic tasks manager actually. So in here, the `RetryTaskMap` is managing the processing regular file tasks.